src/tutorial/index.rst

```diff
@@ -350,66 +343,60 @@ Additionally, we will add the two constants we defined earlier so that they are
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 Now that we've defined the messages, agents and environment for the Circles model, it's time to implement the behaviours of our agents and make use of them.
 
-In FLAME GPU 2, agent functions are implemented using the :c:macro:`FLAMEGPU_AGENT_FUNCTION(name, input_message, output_message)<FLAMEGPU_AGENT_FUNCTION>` macro function. It is expanded by the compiler, to produce the full definition of an agent function (see it's API documentation for an example of it's expansion). However, for our usage we simply need to provide it three parameters; the function's name, the function's message input style and the function's message output style. Then the function can be implemented from this, with the macro call being treated as the function's prototype.
 
-We will start by implementing the agent function, whereby each agent outputs a message sharing their location.
+In FLAME GPU 2, agent functions can be implemented using the C++ :c:macro:`FLAMEGPU_AGENT_FUNCTION(name, input_message, output_message)<FLAMEGPU_AGENT_FUNCTION>` macro function. It is expanded by the compiler, to produce the full definition of an agent function (see it's API documentation for an example of it's expansion). However, for our usage we simply need to provide it three parameters; the function's name, the function's message input style and the function's message output style. Then the function can be implemented from this, with the macro call being treated as the function's prototype.
 
-We will name the function ``output_message`` (the name should not be wrapped in quotes), it does not have a message input so :class:`flamegpu::MessageNone` is used for the input message argument and we're outputting the spatial 2D message we defined above so :class:`flamegpu::MessageSpatial2D` is used for the output message argument.
+The C++ format of agent function description can be compiled at runtime by specifying the function as a C++ string. This enables models specified in Python to compile on the fly. Runtime compilation adds a small additional cost to the initial execution of an agent function, due to compilation. However, FLAME GPU caches compiled agent functions to remove this for repeated runs (if the agent function/model has not changed). When using Python it is possible to specify agent functions using the C++ format as well as via a *pure* Python (a subset of Python referred to as *Agent Python*) description which is shown in this tutorial. Agent function in *pure* Python must be defined as having a ``@pyflamegpu.agent_function`` decorator and using the following syntax ``def outputdata(message_in: pyflamegpu.MessageNone, message_out: pyflamegpu.MessageNone):`` which includes the specifiation of the name and type (uysing type annotations) of the output and input message. The *pure* Python implementation will translate the Python to C++ at runtime prior to compilation through a process known as transpiling.
```

Comment:
It might be nicer to use a multi-line code block for the annotated function definition rather than two inline definiations. 

Also mentions `pure python` still.

Reply:
It breaks up the flow of the description though. The multi line version is in the full agent function example so I'd prefer to leave this as is.